sklearn/linear_model/_cd_fast.pyx

```diff
@@ -180,42 +242,74 @@ def enet_coordinate_descent(
         # tol *= np.dot(y, y)
         tol *= _dot(n_samples, &y[0], 1, &y[0], 1)
 
+        # Check convergence before entering the main loop.
+        gap, dual_norm_XtA = gap_enet(
+            n_samples, n_features, w, alpha, beta, X, y, R, XtA, positive
+        )
+        if gap < tol:
+            with gil:
+                return np.asarray(w), gap, tol, 0
+
+        # Gap Safe Screening Rules, see https://arxiv.org/abs/1802.07481, Eq. 11
+        if do_screening:
+            n_active = 0
+            for j in range(n_features):
+                if norm2_cols_X[j] == 0:
+                    w[j] = 0
+                    excluded_set[j] = 1
+                    continue
+                Xj_theta = XtA[j] / fmax(alpha, dual_norm_XtA)  # X[:,j] @ dual_theta
+                d_j = (1 - fabs(Xj_theta)) / sqrt(norm2_cols_X[j] + beta)
+                if d_j <= sqrt(2 * gap) / alpha:
+                    # include feature j
+                    active_set[n_active] = j
+                    excluded_set[j] = 0
+                    n_active += 1
+                else:
+                    # R += w_j * X[:,j]
+                    _axpy(n_samples, w[j], &X[0, j], 1, &R[0], 1)
+                    w[j] = 0
+                    excluded_set[j] = 1
+
```

Comment:
Would it be possible to extract this code into an inline cdef function wrapped by a private cpdef for the purpose of writing a unit test that checks that the computed excluded set matches the expectations on some toy data?

Reply:
I think the test you suggest would be great indeed.